extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/YoutubeParsingHelper.java

```diff
@@ -1017,80 +1159,176 @@ public static JsonArray getJsonResponse(@Nonnull final Page page,
+                        // Some parameters which are sent by the official WEB client (probably some
+                        // of them are not useful)
+                        .value("currentUrl", "/watch?v=" + videoId)
+                        .value("vis", 0)
+                        .value("splay", false)
+                        .value("autoCaptionsDefaultOn", false)
+                        .value("autonavState", "STATE_NONE")
+                        .value("html5Preference", "HTML5_PREF_WANTS")
+                        .value("signatureTimestamp", sts)
+                        .value("referer", "https://www.youtube.com/watch?v=" + videoId)
+                        .value("lactMilliseconds", "-1")
+                    .end()
+                .end()
+                .value(CPN, contentPlaybackNonce)
+                .value(VIDEO_ID, videoId)
+                .done())
+                .getBytes(StandardCharsets.UTF_8);
+        // @formatter:on
+    }
+
+    /**
+     * Get the user-agent string used as the user-agent for InnerTube requests with the Android
+     * client.
+     *
+     * If the {@link Localization} provided is {@code null}, fallbacks to
+     * {@link Localization#DEFAULT the default one}.
+     *
+     * @param localization the {@link Localization} to set in the user-agent
+     * @return the Android user-agent used for InnerTube requests with the Android client,
+     * depending on the {@link Localization} provided
+     */
+    @Nonnull
+    public static String getAndroidUserAgent(@Nullable final Localization localization) {
+        // Spoofing an Android 12 device with the hardcoded version of the Android app
+        return "com.google.android.youtube/" + MOBILE_YOUTUBE_CLIENT_VERSION
+                + " (Linux; U; Android 12; "
+                + (localization != null ? localization.getCountryCode()
+                        : Localization.DEFAULT.getCountryCode())
+                + ") gzip";
+    }
+
+    /**
+     * Get the user-agent string used as the user-agent for InnerTube requests with the iOS
+     * client.
+     *
+     * If the {@link Localization} provided is {@code null}, fallbacks to
+     * {@link Localization#DEFAULT the default one}.
+     *
+     * @param localization the {@link Localization} to set in the user-agent
+     * @return the iOS user-agent used for InnerTube requests with the iOS client, depending on the
+     * {@link Localization} provided
+     */
+    @Nonnull
+    public static String getIosUserAgent(@Nullable final Localization localization) {
+        // Spoofing an iPhone running iOS 15.4 with the hardcoded mobile client version
+        return "com.google.ios.youtube/" + MOBILE_YOUTUBE_CLIENT_VERSION
+                + "(" + IOS_DEVICE_MODEL
+                + "; U; CPU iOS 15_4 like Mac OS X; "
+                + (localization != null ? localization.getCountryCode()
+                        : Localization.DEFAULT.getCountryCode())
+                + ")";
```

Comment:
The code here (~line 1000 until here) is heavily duplicated (e.g.  ``.value("lockedSafetyMode", false)``) and can many things should be moved into constants. I also don't think that this is a "Helper" class anymore regarding that it has nearly 1700 LoC.
I think key aspects - e.g. the building of the body - should be broken down into subclasses that are inheritable (e.g. a ``Abstract/BaseYTBodyBuilder``,``YTDesktopBodyBuilder``, ``YTAndroid...``)

Reply:
Indeed, but let's not do this in this PR (at least I will not do).